pkg/logbridge/logbridge.go

```diff
@@ -0,0 +1,88 @@
+package logbridge
+
+import (
+	"bufio"
+	"io"
+
+	"github.com/square/p2/pkg/logging"
+)
+
+// This type is agnostic of pipes. os.File conforms to io.Reader and io.Writer
+// so it should not be difficult to migrate
+type Bridge struct {
+	src  io.Reader
+	dest io.Writer
+
+	lines         chan []byte
+	needsTeardown bool
+}
+
+// New takes two files src and dest and returns a composite object of the
+// two.
+func New(src io.Reader, dest io.Writer, capacity int) Bridge {
+	if src == nil {
+		panic("Argument Error: Reader cannot be nil")
+	}
+	if dest == nil {
+		panic("Argument Error: Reader cannot be nil")
+	}
+
+	return Bridge{
+		src:           src,
+		dest:          dest,
+		lines:         make(chan []byte, capacity),
+		needsTeardown: true,
+	}
+}
+
+// Run is responsible for reading from src and writing to dst until src is
+// consumed or an error is encountered. The caller is responsible for closing
+// and flushing the reader and writer.
+func (b *Bridge) Run() error {
```

Comment:
One thing I regret about the API of farms which you might want to consider here: if Run and New are separate actions, then it should be safe to call Run repeatedly on a single bridge or otherwise enforce invariants so that it can't be done. I'm not sure it is in this design (since the lines channel is scoped to the structure and not to an instance of Run).

In this case I don't see much value to separating the two functions. I prefer the design of ConsulSessionManager where there is no struct at all and the function just spins out a goroutine.